src/Umbraco.Commerce.PaymentProviders.Quickpay/QuickpayCheckoutPaymentProvider.cs

```diff
@@ -172,66 +174,74 @@ public override async Task<PaymentFormResult> GenerateFormAsync(PaymentProviderC
+                    return CallbackResult.BadRequest();
+                }
+
+                QuickpayPayment payment = await ParseCallbackAsync(
+                        context.Request,
                         cancellationToken).ConfigureAwait(false);
 
-                    if (VerifyOrder(ctx.Order, payment))
-                    {
-                        // Get operations to check if payment has been approved
-                        var operation = payment.Operations.LastOrDefault();
+                if (!VerifyOrder(context.Order, payment))
+                {
+                    Logger.Warn($"Quickpay [{context.Order.OrderNumber}] - Couldn't verify the order");
+                    return CallbackResult.Empty;
+                }
 
-                        // Check if payment has been approved
-                        if (operation != null)
-                        {
-                            var totalAmount = operation.Amount;
+                Operation latestOperation = payment.Operations.LastOrDefault();
+                if (latestOperation == null)
+                {
+                    return CallbackResult.BadRequest();
+                }
 
-                            if (operation.QuickpayStatusCode == "20000" || operation.AcquirerStatusCode == "000")
-                            {
-                                var paymentStatus = GetPaymentStatus(operation);
-
-                                return new CallbackResult
-                                {
-                                    TransactionInfo = new TransactionInfo
-                                    {
-                                        AmountAuthorized = AmountFromMinorUnits(totalAmount),
-                                        TransactionId = GetTransactionId(payment),
-                                        PaymentStatus = paymentStatus
-                                    }
-                                };
-                            }
-                            else
-                            {
-                                Logger.Warn($"Quickpay [{ctx.Order.OrderNumber}] - Payment not approved. Quickpay status code: {operation.QuickpayStatusCode} ({operation.QuickpayStatusMessage}). Acquirer status code: {operation.AcquirerStatusCode} ({operation.AcquirerStatusMessage}).");
-                            }
-                        }
+                if (latestOperation.QuickpayStatusCode == QuickpayStatusCodeApproved || latestOperation.AcquirerStatusCode == "000")
+                {
+                    PaymentStatus? currentPaymentStatus = context.Order?.TransactionInfo?.PaymentStatus;
+                    PaymentStatus newPaymentStatus = GetPaymentStatus(latestOperation);
+
+                    Logger.Debug($"ProcessCallbackAsync - current payment status: {currentPaymentStatus}, new payment status: {newPaymentStatus}, operations: {System.Text.Json.JsonSerializer.Serialize(payment.Operations)}.");
```

Comment:
I think it would be best to use same JSON serializer, so either `Newtonsoft.Json` ... or better replace with `System.Text.Json` (but probably in a minor version - just in case some serialization is different).

Reply:
yeah, I'm working on a version that migrate from `newtonsoft.json` to `sytem.text.json` but it's planned for v13.1.0 instead. This one should be a patch version.